pkg/controller/certificatesigningrequest/ipsec_csr_signing_controller.go

```diff
@@ -0,0 +1,486 @@
+	if err != nil {
+		return err
+	}
+	currCA, ok := c.certificateAuthority.Load().(*certificateAuthority)
+	if !ok || currCA == nil {
+		return fmt.Errorf("certificate authority is not initialized")
+	}
+	signed, err := currCA.signCSR(template, req.PublicKey)
+	if err != nil {
+		return err
+	}
+	bs, err := certutil.EncodeCertificates(signed)
+	if err != nil {
+		return err
+	}
+	toUpdate := csr.DeepCopy()
+	toUpdate.Status.Certificate = bs
+	_, err = c.client.CertificatesV1().CertificateSigningRequests().UpdateStatus(context.TODO(), toUpdate, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func newCertificateTemplate(certReq *x509.CertificateRequest, usage []certificatesv1.KeyUsage) (*x509.Certificate, error) {
+	var sn big.Int
+	snBytes := make([]byte, 18)
+	_, err := rand.Read(snBytes)
+	if err != nil {
+		return nil, err
+	}
+	sn.SetBytes(snBytes)
+	keyUsage, extKeyUsage, err := keyUsagesFromStrings(usage)
+	if err != nil {
+		return nil, err
+	}
+	template := &x509.Certificate{
+		Subject:               certReq.Subject,
+		SignatureAlgorithm:    x509.SHA512WithRSA,
+		NotBefore:             time.Now().Add(-5 * time.Minute),
+		NotAfter:              time.Now().Add(duration365d), // defaults to 1 year
+		SerialNumber:          &sn,
+		DNSNames:              certReq.DNSNames,
+		BasicConstraintsValid: true,
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           extKeyUsage,
+	}
+	return template, nil
+}
+
+func (c *IPsecCSRSigningController) processNextFixtureWorkItem() bool {
+	key, quit := c.fixturesQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.fixturesQueue.Done(key)
+	err := c.syncRootCertificateAndKey()
+	if err != nil {
+		c.fixturesQueue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync root CA and private key")
+		return true
+	}
+	c.fixturesQueue.Forget(key)
+	return true
+}
+
+func (c *IPsecCSRSigningController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.syncCSR(key.(string))
+	if err != nil {
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Failed to sync CertificateSigningRequest", "CertificateSigningRequest", key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+// generateSelfSignedRootCertificate creates self-signed CA certificates and returns the PEM encoded
+// certificates and private key.
+func generateSelfSignedRootCertificate(commonName string) ([]byte, []byte, error) {
+	validFrom := time.Now().Add(-time.Hour) // valid an hour earlier to avoid flakes due to clock skew
+	caKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+	caTemplate := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName:   commonName,
+			Organization: []string{antreaapis.AntreaOrganizationName},
+		},
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(duration10y),
```

Comment:
Most self-signed certificate sets 1 year duration, I'm not sure the risk but could it keep align?
Even if this is 10 year, how it handles renewal? Maybe we could skip CA renewal in this version, but the duration should be conservative.

Reply:
10 years cert is for CA only. Do you think 1 year is enough for CA? Then what would be the proper expiration period for signed certificates? I saw kubeadm by default creates 10-year CA for clusters and does not support rotating CA automatically as well.

Reply:
OK, do you know how kubeadm replaces CA if it's leaked? For antrea, if deleting the secret and restarting all antrea components can fix it, it works for me.

Reply:
kubernetes provides [manual steps](https://kubernetes.io/docs/tasks/tls/manual-rotation-of-ca-certificates/) to rotate CA. And for our case, we just need to delete the secret and restart Agent Pods.